### PR TITLE
Add AWS configuration hook called by post-startup.sh.

### PR DESCRIPTION
**Pull request by jmczerk**
This ports over the configuration that we do in the SageMaker startup scripts to help automate access to Workbench resources from AWS compute.  This will get pared down quite a bit once we implement credential-less AuthN for AWS instances, but that's at least a couple sprints away.

### CODE REVIEW

File: startupscript/aws/post-startup-hook.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+# Set up aws-vault credential caching
+#####################################
+${RUN_AS_LOGIN_USER} "${WORKBENCH_INSTALL_PATH} config set cache-with-aws-vault true"
+${RUN_AS_LOGIN_USER} "${WORKBENCH_INSTALL_PATH} config set wb-path --path ${WORKBENCH_INSTALL_PATH}"
+${RUN_AS_LOGIN_USER} "${WORKBENCH_INSTALL_PATH} config set aws-vault-path --path ${AWS_VAULT_INSTALL_PATH}"
+
+###############################################################################
+# Write common environment vars to .profile so they are present in all contexts
+###############################################################################
+readonly WORKBENCH_WORKSPACE=$(get_metadata_value terra-workspace-id)
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:44]:
```suggestion
readonly WORKBENCH_WORKSPACE="$(get_metadata_value terra-workspace-id)"
```

File: startupscript/aws/post-startup-hook.sh [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+readonly AWS_CONFIG_FILE="${USER_WORKBENCH_CONFIG_DIR}/aws/${WORKBENCH_WORKSPACE}.conf"
+cat >> "${USER_PROFILE}" << EOM
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:46]: Can we `s/EOM/EOF` as we've previously used `EOF`?

File: startupscript/aws/post-startup-hook.sh [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+export AWS_VAULT_FILE_PASSPHRASE=""
+EOM
+chown "${USER_PRIMARY_GROUP}:${USER_NAME}" "${USER_PROFILE}"
+
+#######################################
+# Write VWB helper functions to .bashrc
+#######################################
+cat >> "${USER_BASHRC}" << 'EOM'
+
+configure_workspace() {
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:62]: Have encouraged the pattern for functions:

```
function my_fn() {
...
}
readonly -f my_fn
```

Slightly more verbose, but makes the code a bit more readable and marking functions readonly prevents an unpleasant category of bugs from cropping up.

File: startupscript/aws/post-startup-hook.sh [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  ${WORKBENCH_INSTALL_PATH} workspace set --uuid ${WORKBENCH_WORKSPACE}
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:63]:
```suggestion
  "${WORKBENCH_INSTALL_PATH}" workspace set --uuid "${WORKBENCH_WORKSPACE}"
```

File: startupscript/aws/post-startup-hook.sh [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  ${WORKBENCH_INSTALL_PATH} workspace configure-aws --cache-with-aws-vault=true
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:64]:
```suggestion
  "${WORKBENCH_INSTALL_PATH}" workspace configure-aws --cache-with-aws-vault=true
```

File: startupscript/aws/post-startup-hook.sh [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  ${WORKBENCH_INSTALL_PATH} resource mount
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:65]:
```suggestion
  "${WORKBENCH_INSTALL_PATH}" resource mount
```

File: startupscript/aws/post-startup-hook.sh [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+configure_ssh() {
+  local USER_SSH_DIR="${HOME}/.ssh"
+  mkdir -p ${USER_SSH_DIR}
+  local USER_SSH_KEY=$(${WORKBENCH_INSTALL_PATH} security ssh-key get --include-private-key --format=JSON)
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:71]:
```suggestion
  local USER_SSH_KEY="$("${WORKBENCH_INSTALL_PATH}" security ssh-key get --include-private-key --format=JSON)"
```

File: startupscript/aws/post-startup-hook.sh [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  echo $USER_SSH_KEY | jq -r '.privateSshKey' > ${USER_SSH_DIR}/id_rsa
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:72]:
```suggestion
  echo "${USER_SSH_KEY}" | jq -r '.privateSshKey' > "${USER_SSH_DIR}"/id_rsa
```

File: startupscript/aws/post-startup-hook.sh [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  echo $USER_SSH_KEY | jq -r '.publicSshKey' > ${USER_SSH_DIR}/id_rsa.pub
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:73]:
```suggestion
  echo "${USER_SSH_KEY}" | jq -r '.publicSshKey' > "${USER_SSH_DIR}"/id_rsa.pub
```

File: startupscript/aws/post-startup-hook.sh [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  chmod 0600 ${USER_SSH_DIR}/id_rsa*
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:74]:
```suggestion
  chmod 0600 "${USER_SSH_DIR}"/id_rsa*
```

File: startupscript/aws/post-startup-hook.sh [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+  ssh-keyscan -H github.com >> ${USER_SSH_DIR}/known_hosts
+}
+
+configure_git() {
+  mkdir -p "${WORKBENCH_GIT_REPOS_DIR}"
+  pushd "${WORKBENCH_GIT_REPOS_DIR}"
+  ${WORKBENCH_INSTALL_PATH} resource list --type=GIT_REPO --format json | jq -c .[] | while read ITEM; do
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:81]: Reformat a little to make it more legible:

```
  "${WORKBENCH_INSTALL_PATH}" resource list --type=GIT_REPO --format json | \
    jq -c .[] | \
    while read ITEM; do
      local GIT_REPO_NAME="$(echo $ITEM | jq -r .id)"
      local GIT_REPO_URL="$(echo $ITEM | jq -r .gitRepoUrl)"
      if [[ ! -d "${GIT_REPO_NAME}" ]]; then
        git clone "${GIT_REPO_URL}" "${GIT_REPO_NAME}"
      fi
    done
```

File: startupscript/aws/post-startup-hook.sh [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+  configure_git
+}
+EOM
+
+##################################
+# Write login prompt .bash_profile
+##################################
+cat >> "${USER_BASH_PROFILE}" << 'EOM'
+
+if [ $(${WORKBENCH_INSTALL_PATH} auth status --format json | jq .loggedIn) == false ]; then
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:103]:
```suggestion
if [[ "$("${WORKBENCH_INSTALL_PATH}" auth status --format json | jq .loggedIn)" == false ]]; then
```

File: startupscript/aws/post-startup-hook.sh [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    echo "User must log into Workbench to continue."
+    ${WORKBENCH_INSTALL_PATH} auth login
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:105]:
```suggestion
    "${WORKBENCH_INSTALL_PATH}" auth login
```

File: startupscript/post-startup.sh [hunk 14]
```diff
@@ -19,22 +19,25 @@ readonly LOG_IN="${4}"
...
 # Emit a message with a timestamp
 #######################################
 source ${SCRIPT_DIR}/emit.sh
 
-source ${SCRIPT_DIR}/${CLOUD}/vm-metadata.sh
+source ${CLOUD_SCRIPT_DIR}/vm-metadata.sh
 
 readonly RUN_AS_LOGIN_USER="sudo -u ${USER_NAME} bash -l -c"
 
+readonly USER_PRIMARY_GROUP="$(id --group --name ${USER_NAME})"
```
Comment (mbookman) [inline startupscript/post-startup.sh:32]:
```suggestion
readonly USER_PRIMARY_GROUP="$(id --group --name "${USER_NAME}")"
```

File: startupscript/post-startup.sh [hunk 15]
```diff
@@ -19,22 +19,25 @@ readonly LOG_IN="${4}"
...
 # absolute path is more reliable.
 readonly SCRIPT_DIR="$(dirname "$(readlink -f "$0")")"
+readonly CLOUD_SCRIPT_DIR="${SCRIPT_DIR}/${CLOUD}"
 #######################################
 # Emit a message with a timestamp
 #######################################
 source ${SCRIPT_DIR}/emit.sh
 
-source ${SCRIPT_DIR}/${CLOUD}/vm-metadata.sh
+source ${CLOUD_SCRIPT_DIR}/vm-metadata.sh
```
Comment (mbookman) [inline startupscript/post-startup.sh:28]:
```suggestion
source "${CLOUD_SCRIPT_DIR}"/vm-metadata.sh
```

File: startupscript/post-startup.sh [hunk 16]
```diff
@@ -112,4 +120,11 @@ fi
 #############################
 # Mount buckets
 #############################
-source ${SCRIPT_DIR}/${CLOUD}/resource-mount.sh
+source ${CLOUD_SCRIPT_DIR}/resource-mount.sh
```
Comment (mbookman) [inline startupscript/post-startup.sh:123]:
```suggestion
source "${CLOUD_SCRIPT_DIR}"/resource-mount.sh
```

**Review by mbookman — COMMENTED** [review]
LG. Just a few things...

**Comment by jmczerk** [comment]
Thank you for the thorough review @mbookman !  I especially appreciate you adding the suggested changes.

I've been trying to use `shellcheck` in VSCode locally to try to be cleaner, but this doesn't help with heredoc embedded code...  I'm considering maybe breaking these kind of things into separate files so they can be linted, but that also might clutter the directory a bit (and I'm guessing the import syntax might be ugly?)

I am also considering adding a `shellcheck` PR check workflow using something like [this](https://github.com/ludeeus/action-shellcheck).

WDYT?

**Comment by mbookman** [comment]
> I've been trying to use shellcheck in VSCode locally to try to be cleaner, but this doesn't help with heredoc embedded code... I'm considering maybe breaking these kind of things into separate files so they can be linted, but that also might clutter the directory a bit (and I'm guessing the import syntax might be ugly?)

If we can establish clear patterns (a little work on file/directory layout and some documentation) for maintainers to follow, I am a fan of using off-the-shelf and custom tooling for better code hygiene. Getting 70% of the way there is usually just annoying enough to make it harder to maintain and getting 90% of the way there gets you the benefits without the cost of dotting all the i's and crossing all the t's.

**Comment by jmczerk** [comment]
@mbookman I ended up breaking out the literal heredocs that get appended to `.bashrc` and `.bash_profile` out into separate files so that they will be linted with `shellcheck` when https://github.com/verily-src/workbench-app-devcontainers/pull/67 goes in.  PTAL and approve if you're satisfied (I'd like to get this in ahead of the shellcheck PR).

File: startupscript/aws/post-startup-hook.sh [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+# - WORKBENCH_INSTALL_PATH: path to CLI executable
+
+readonly AWS_VAULT_INSTALL_PATH="/usr/bin/aws-vault"
+
+##########################################
+# Install aws-vault for credential caching
+##########################################
+emit "installing aws-vault"
+curl --no-progress-meter --location --output "${AWS_VAULT_INSTALL_PATH}" \
+    https://github.com/99designs/aws-vault/releases/download/v7.2.0/aws-vault-linux-amd64
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:32]: It's a nit, but I think this URL would be good as a variable defined up with the install path.

File: startupscript/aws/post-startup-hook.sh [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+${RUN_AS_LOGIN_USER} "${WORKBENCH_INSTALL_PATH} config set cache-with-aws-vault true"
+${RUN_AS_LOGIN_USER} "${WORKBENCH_INSTALL_PATH} config set wb-path --path ${WORKBENCH_INSTALL_PATH}"
+${RUN_AS_LOGIN_USER} "${WORKBENCH_INSTALL_PATH} config set aws-vault-path --path ${AWS_VAULT_INSTALL_PATH}"
+
+###############################################################################
+# Write common environment vars to .profile so they are present in all contexts
+###############################################################################
+WORKBENCH_WORKSPACE="$(get_metadata_value terra-workspace-id)"
+AWS_CONFIG_FILE="${USER_WORKBENCH_CONFIG_DIR}/aws/${WORKBENCH_WORKSPACE}.conf"
+readonly WORKBENCH_WORKSPACE AWS_CONFIG_FILE
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:47]: I'd like for us to have the pattern of one variable per line. I don't love using more vertical space, but I think it reads easier.

```suggestion
readonly WORKBENCH_WORKSPACE
readonly AWS_CONFIG_FILE

```

File: startupscript/aws/post-startup-hook.sh [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+#######################################
+# Write VWB helper functions to .bashrc
+#######################################
+cat "${CLOUD_SCRIPT_DIR}/bashrc-functions.bash" >> "${USER_BASHRC}"
+
+##################################
+# Write login prompt .bash_profile
+##################################
+cat "${CLOUD_SCRIPT_DIR}/bash-profile-append.bash" >> "${USER_BASH_PROFILE}"
```
Comment (mbookman) [inline startupscript/aws/post-startup-hook.sh:67]: Missing EOL.

File: startupscript/post-startup.sh [hunk 20]
```diff
@@ -112,4 +120,11 @@ fi
...
 # Mount buckets
 #############################
-source ${SCRIPT_DIR}/${CLOUD}/resource-mount.sh
+source "${CLOUD_SCRIPT_DIR}"/resource-mount.sh
+
+###############################
+# cloud platform specific setup
+###############################
+if [[ -f "${CLOUD_SCRIPT_DIR}/post-startup-hook.sh" ]]; then
+    source "${CLOUD_SCRIPT_DIR}/post-startup-hook.sh"
```
Comment (mbookman) [inline startupscript/post-startup.sh:129]:
```suggestion
  source "${CLOUD_SCRIPT_DIR}/post-startup-hook.sh"
```

**Review by mbookman — CHANGES_REQUESTED** [review]

File: startupscript/post-startup.sh [hunk 21]
```diff
@@ -54,20 +57,31 @@ exec 2>&1
...
+# script writes to  ~/.bashrc, and has the ~/.bash_profile source the ~/.bashrc
+
 cat << EOF >> "${USER_BASH_PROFILE}"
 
 if [[ -e ~/.bashrc ]]; then
  source ~/.bashrc
 fi
 
 EOF
+chown "${USER_NAME}:${USER_PRIMARY_GROUP}" "${USER_BASH_PROFILE}"
```
Comment (yuhuyoyo) [inline startupscript/post-startup.sh:84]: do you mind fixing it here too? https://github.com/verily-src/workbench-app-devcontainers/blob/master/startupscript/install-java.sh#L16

**Review by yuhuyoyo — APPROVED** [review]

**Review by mbookman — APPROVED** [review]
LGTM

**Pull request merged**